### ISSUE TITLE
learningRate seted when use single or multi GPU

### ISSUE THREAD

**Issue by buaacarzp** [BIGBALLON/distribuuuu#15]
your learning rate isn't modified when changing the single gpu to multi gpu, are  you noticed it ?

**Comment by BIGBALLON** [comment]
Hi,  @buaacarzp 

You can see the baseline for LR settings(try to check pycls [Model Zoo](https://github.com/facebookresearch/pycls/blob/master/MODEL_ZOO.md) for more details):

I use a **reference** learning rate of **0.1(batch size=128)** and a weight decay of **5e-5**,
so the **LR** is set to **6.4**  when the batch size is **8192** (128*64GPUs), check the following results:

|                    model                     | epoch |    total batch     | lr policy | base lr | Acc@1  | Acc@5  |                                                       model / config                                                        |
| :------------------------------------------: | :---: | :----------------: | :-------: | :-----: | :----: | :----: | :-------------------------------------------------------------------------------------------------------------------------: |
|                   resnet18                   |  100  |   256 (32*8GPUs)   |    cos    |   0.2   | 70.902 | 89.894 | [Drive](https://drive.google.com/file/d/18a6QFc_DoTHo3TWkN_EsptyGmhF97sVw/view?usp=sharing) / [cfg](./config/resnet18.yaml) |
|                   resnet18                   |  100  |  1024 (128*8GPUs)  |    cos    |   0.8   | 70.994 | 89.892 |                                                                                                                             |
|                   resnet18                   |  100  | 8192 (128*64GPUs)  |    cos    |   6.4   | 70.165 | 89.374 |                                                                                                                             |
|                   resnet18                   |  100  | 16384 (256*64GPUs) |    cos    |  12.8   | 68.766 | 88.381 |                                                                                                                             |
|                   resnet50                   |  100  |   256 (32*8GPUs)   |    cos    |   0.2   | 77.252 | 93.430 | [Drive](https://drive.google.com/file/d/1rUY1mSYTxe7jWzzcWrreg398tbSNXtnv/view?usp=sharing) / [cfg](./config/resnet50.yaml) |
| [botnet50](https://arxiv.org/abs/2101.11605) |  100  |   256 (32*8GPUs)   |    cos    |   0.2   | 77.604 | 93.682 | [Drive](https://drive.google.com/file/d/1-jvhJaMyy-KziAuFnmt5rkoZrm5364UF/view?usp=sharing) / [cfg](./config/botnet50.yaml) |
|                  resnext101                  |  100  |   256 (32*8GPUs)   |    cos    |   0.2   | 78.938 | 94.482 |

**Comment by BIGBALLON** [comment]
TODO:

- clarify the training settings
- add NUM_CLASSES 
- add timm
- add more baseline

**Comment by buaacarzp** [comment]
> Hi, @buaacarzp
> 
> You can see the baseline for LR settings (try to check pycls [Model Zoo](https://github.com/facebookresearch/pycls/blob/master/MODEL_ZOO.md) for more details ):
> 
> I use a **reference** learning rate of **0.1(batch size=128)** and a weight decay of **5e-5**,
> so the **LR** is set to **6.4** when the batch size is **8192** (128*64GPUs), check the following results:
> 
> model	epoch	total batch	lr policy	base lr	Acc@1	Acc@5	model / config
> resnet18	100	256 (32*8GPUs)	cos	0.2	70.902	89.894	[Drive](https://drive.google.com/file/d/18a6QFc_DoTHo3TWkN_EsptyGmhF97sVw/view?usp=sharing) / [cfg](./config/resnet18.yaml)
> resnet18	100	1024 (128*8GPUs)	cos	0.8	70.994	89.892	
> resnet18	100	8192 (128*64GPUs)	cos	6.4	70.165	89.374	
> resnet18	100	16384 (256*64GPUs)	cos	12.8	68.766	88.381	
> resnet50	100	256 (32*8GPUs)	cos	0.2	77.252	93.430	[Drive](https://drive.google.com/file/d/1rUY1mSYTxe7jWzzcWrreg398tbSNXtnv/view?usp=sharing) / [cfg](./config/resnet50.yaml)
> [botnet50](https://arxiv.org/abs/2101.11605)	100	256 (32*8GPUs)	cos	0.2	77.604	93.682	[Drive](https://drive.google.com/file/d/1-jvhJaMyy-KziAuFnmt5rkoZrm5364UF/view?usp=sharing) / [cfg](./config/botnet50.yaml)
> resnext101	100	256 (32*8GPUs)	cos	0.2	78.938	94.482

Okay, here's a question:
1. Why is the single-server multi-card training speed of the `mp` slower than the `launch` time?
2. Why is the time taken on the CIFAR10 server slower than the time used on the `launch` single-device multi-card server?
3. Why is the CUDA memory usage sometimes 0% when multiple GPU cards are used on a single device during data set training? However, the CUDA memory usage of both GPU cards on the CIFAR10 is 95% +?

**Comment by BIGBALLON** [comment]
The biggest question: IO time
Something IO is the real bottleneck, not GPU.

CIFAR10 is loaded in memory. If you have a server with large RAM(such as 512GB), you can also loaded whole imagenet dataset. 

BTW, I think there is no essential difference between mp and launch..XD

**Comment by BIGBALLON** [comment]
FYI 
https://www.zhihu.com/answer/907835663

**Comment by buaacarzp** [comment]
> The biggest question: IO time
> Something IO is the real bottleneck, not GPU.
> 
> CIFAR10 is loaded in memory. If you have a server with large RAM(such as 512GB), you can also loaded whole imagenet dataset.
> 
> BTW, I think there is no essential difference between mp and launch..XD

Hi,Bro,You mean the dataloader in CIFAR 10 has pin_memory set as beblow?
```
train_loader = torch.utils.data.DataLoader(
        trainset,
        batch_size=BATCH_SIZE,
        num_workers=4,
        pin_memory=True,
        sampler=train_sampler,
    )
```

**Comment by BIGBALLON** [comment]
I mean try to focus on ImageNet, not CIFAR

**Comment by buaacarzp** [comment]
But you say that `CIFAR10 is loaded in memory. `, and then you said try it on imagenet. The problem I have now is that the dataloader reads data slowly. It takes 3 seconds on average to read data in each batch. Is this related to my data? Because my data is video data, I store it as an npy file, read the image and save it as an npy file. Why?

**Comment by BIGBALLON** [comment]
1: ``CIFAR10 is loaded in memory`` means **the whole dataset(50K train images and 10K test images)** are loaded in memory, so the IO is faster than ImageNet.
2: ``Is this related to my data``, I think so.
3: ``pin_memory`` 's help is limited
4: ``Why?`` check your dataset and code...XD.

**Comment by BIGBALLON** [comment]
5: The problem I have now is that the dataloader reads data slowly. It takes 3 seconds on average to read data in each batch. 

This is related to your hardware device(CPU, SSD) and the dataset itself

**Comment by BIGBALLON** [comment]
6: Is there any preprocessing after loading your video data? try to find the function or position which costs the most time.

**Comment by buaacarzp** [comment]
1.Why is the CIFAR10 data set loaded in memory? Is it because of the `pin_memory` setting? I set it to True when I loaded my dataset, but only one card was 100% during training, and the other card was 0 for a long time and the cuda memory increased occasionally.
2.Another problem: I've been using your dis~uuu library these past two days. I find that when calculating acc, the acc tested only on the split training set or verification set does not represent the whole data set. Is there any document to look at to solve this problem?

**Comment by BIGBALLON** [comment]
> 1.Why is the CIFAR10 data set loaded in memory? Is it because of the pin_memory setting? I set it to True when I loaded my dataset, but only one card was 100% during training, and the other card was 0 for a long time and the cuda memory increased occasionally.

A: Plz check the following snippets:

https://github.com/pytorch/vision/blob/cac8a97b0bd14eddeff56f87a890d5cc85776e18/torchvision/datasets/cifar.py#L12
https://github.com/pytorch/vision/blob/cac8a97b0bd14eddeff56f87a890d5cc85776e18/torchvision/datasets/cifar.py#L80

You may find why the CIFAR10 data set loaded in memory.

as for pin_memory, plz check this [article](https://pytorch.org/docs/stable/data.html#memory-pinning), For data loading, passing pin_memory=True to a DataLoader will automatically put the fetched data Tensors in pinned memory, and thus enables faster data transfer to CUDA-enabled GPUs. and what is ``pinned memory``, plz review the knowledge of the Operating System.


> 2.Another problem: I've been using your dis~uuu library these past two days. I find that when calculating acc, the acc tested only on the split training set or verification set does not represent the whole data set. Is there any document to look at to solve this problem?

A:  This is the standard test method based on ResNet paper(see this [script](https://gist.github.com/BIGBALLON/8a71d225eff18d88e469e6ea9b39cef4)), I don't think there is any problem.

**Comment by BIGBALLON** [comment]
and see https://github.com/pytorch/vision/blob/cac8a97b0bd14eddeff56f87a890d5cc85776e18/torchvision/datasets/folder.py#L272, check the usage of ``torchvision.datasets.ImageFolder``

**Comment by BIGBALLON** [comment]
Hi, @buaacarzp, I update the settings, see https://github.com/BIGBALLON/distribuuuu/commit/9223904604f8aaedee2ddd128c0ef0023ece24d3

**Comment by buaacarzp** [comment]
> Hi, @buaacarzp, I update the settings, see [9223904](https://github.com/BIGBALLON/distribuuuu/commit/9223904604f8aaedee2ddd128c0ef0023ece24d3)

I saw it in great detail. good job.
